src/transformers/generation/candidate_generator.py

```diff
@@ -262,6 +348,212 @@ def update_candidate_strategy(self, input_ids: torch.LongTensor, scores: torch.F
+        if has_past_key_values:
+            new_cache_size = new_cur_len - 1 - remove_from_pkv
+            self.assistant_kwargs["past_key_values"] = _crop_past_key_values(
+                self.assistant_model, self.assistant_kwargs["past_key_values"], new_cache_size - 1
+            )  # the assistant does not have the token after the last match, hence the -1
+
+            self.assistant_kwargs = _prepare_attention_mask(
+                self.assistant_kwargs, new_cur_len, self.assistant_model.config.is_encoder_decoder
+            )
+            self.assistant_kwargs = _prepare_token_type_ids(self.assistant_kwargs, new_cur_len)
+
+        # 2. Forecast next N tokens using the assistant model.
+        assistant_generation_kwargs = {
+            self.input_ids_key: assistant_input_ids,
+            "min_new_tokens": min_new_tokens,
+            "max_new_tokens": max_new_tokens,
+            "generation_config": self.generation_config,
+            "logits_processor": self.logits_processor,
+        }
+
+        self.assistant_kwargs.pop("attention_mask", None)
+
+        assistant_output = self.assistant_model.generate(**assistant_generation_kwargs, **self.assistant_kwargs)
+
+        num_prev_assistant = self.prev_assistant_ids.shape[1]
+        start_assistant_look_index = num_prev_assistant - self.assistant_lookbehind
+
+        new_target_ids_from_window = self.convert_token_ids(
+            assistant_output.sequences[:, start_assistant_look_index:],
+            src=self.assistant_tokenizer,
+            dest=self.target_tokenizer,
+        )
+        target_prompt_use_length = new_target_ids_from_window.shape[1]
+
+        target_prompt_use = input_ids[:, -target_prompt_use_length:]
+
+        (
+            tar_replace_tokens_from_prompt,
+            tar_disrep_length,
+            tar_new_tokens_with_disrep,
+            tar_new_tokens_only,
+            tar_discrep_only,
+        ) = _get_tokens_diag(target_prompt_use, new_target_ids_from_window)
+
+        new_target_ids = input_ids
+
+        if tar_new_tokens_only is not None:
+            if tar_new_tokens_only.shape[1] > 0:
+                new_target_ids = torch.cat([new_target_ids, tar_new_tokens_only], dim=-1)
+        else:
+            # edge case: in case of no intersection between prompt and new_target_ids
+            new_target_ids = torch.cat([new_target_ids, new_target_ids_from_window], dim=-1)
+
+        self.prev_target_ids = input_ids
+
+        for stopping_criterion in stopping_criteria:
+            if hasattr(stopping_criterion, "max_length"):
+                new_target_ids = new_target_ids[:, : stopping_criteria.max_length]
+
+        # 3. Update variables for the next round of candidate generation
+        self.assistant_kwargs["past_key_values"] = assistant_output.past_key_values
+        self.prev_tokens = assistant_output.sequences
+
+        # 4. Prepare variables for output
+        if input_ids.shape[1] >= new_target_ids.shape[1]:
```

Comment:
In this (corner) case, the assistant model is about to return fewer tokens than it received as input. In the main generation loop, we simply want the main model to ignore the candidate sequences and generate 1 new token.

We can achieve this if we do `return input_ids, None`, as opposed to appending dummy tokens for the main model to reject 🤗 This would be the same as the return above

Reply:
good catch, thanks !